cpp/src/spatial/point_in_polygon.cu

```diff
@@ -0,0 +1,233 @@
+
+template <typename T>
+__global__ void point_in_polygon_kernel(cudf::size_type num_test_points,
+                                        const T* const __restrict__ test_points_x,
+                                        const T* const __restrict__ test_points_y,
+                                        cudf::size_type num_polys,
+                                        const cudf::size_type* const __restrict__ poly_offsets,
+                                        const cudf::size_type* const __restrict__ poly_ring_offsets,
+                                        const T* const __restrict__ poly_points_x,
+                                        const T* const __restrict__ poly_points_y,
+                                        cudf::size_type* const __restrict__ result)
+{
+    cudf::size_type idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    if (idx > num_test_points) {
+        return;
+    }
+
+    cudf::size_type hit_mask = 0;
+
+    T x = test_points_x[idx];
+    T y = test_points_y[idx];
+
+    // for each polygon
+    for (cudf::size_type poly_idx = 0; poly_idx < num_polys; poly_idx++)
+    {
+        cudf::size_type poly_begin = (0 == poly_idx) ? 0 : poly_offsets[poly_idx - 1];
+        cudf::size_type poly_end = poly_offsets[poly_idx];
+        bool point_is_within = false;
+
+        // for each ring
+        for (cudf::size_type ring_idx = poly_begin; ring_idx < poly_end; ring_idx++)
+        {
+            cudf::size_type ring_begin = (ring_idx == 0) ? 0 : poly_ring_offsets[ring_idx - 1];
+            cudf::size_type ring_end = poly_ring_offsets[ring_idx];
+
+            // for each line segment
+            for (cudf::size_type point_idx = ring_begin; point_idx < ring_end - 1; point_idx++)
+            {
+                T ax = poly_points_x[point_idx];
+                T ay = poly_points_y[point_idx];
+                T bx = poly_points_x[point_idx + 1];
+                T by = poly_points_y[point_idx + 1];
+
+                auto y_between_ay_by = ay <= y && y < by; // is y in range [ay, by) when ay < by?
+                auto y_between_by_ay = by <= y && y < ay; // is y in range [by, ay) when by < ay?
+                auto y_in_bounds = y_between_ay_by || y_between_by_ay; // is y in range [by, ay]?
+                auto run  = bx - ax;
+                auto rise = by - ay;
+                auto rise_to_point =  y - ay;
+
+                if (y_in_bounds && x < (run / rise) * rise_to_point + ax) {
+                    point_is_within = not point_is_within;
+                }
+            }
+        }
+
+        hit_mask |= point_is_within << poly_idx;
+    }
+
+    result[idx] = hit_mask;
+}
+
+struct point_in_polygon_functor
+{
+    template <typename T>
+    static constexpr bool is_supported()
+    {
+        return std::is_floating_point<T>::value;
+    }
+
+    template <typename T, std::enable_if_t< !is_supported<T>() >* = nullptr, typename... Args>
+    std::unique_ptr<cudf::column> operator()(Args&& ...)
```

Comment:
This is a nice way to not repeat all the args in the unsupported path, I'll have to use this idea.

Reply:
Can't take credit for that one!

https://github.com/rapidsai/cudf/pull/3402#discussion_r358425449